Add support for restricting the secrets watch list in cainjector

aubm opened the pull request:
The --secrets-field-selector flag is introduced and accepts a field selector expression

The following example makes cainjector ignore all secrets whose type is "helm.sh/release.v1".

```
--secrets-field-selector type!=helm.sh/release.v1
```

<!--

Thanks for opening a pull request! Here are some tips to get everything merged smoothly:

1. Read our contributor guidelines: https://cert-manager.io/docs/contributing/

2. Make sure your commits are signed off: https://cert-manager.io/docs/contributing/sign-off/

3. If the PR is unfinished, raise it as a draft or prefix the title with "WIP:" so it's clear to everyone.

4. Be sure to allow edits from maintainers so it's easier for us to help: https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork

-->

### Pull Request Motivation

<!-- Explain the motivation behind this PR. If there's a related issue or PR, link to it here! -->

Some clusters have lots of creations/updates of secrets that can contain a significant amount of data, namely when heavily relying on helm 3 to deploy lots of releases.
In these clusters, the cainjector can see its memory consumption occasionaly peaks and eventually OOM crash when the allocated limit is exceeded.

To avoid that behavior, one could levarage the `--secrets-label-selector` flags introduced in the PR in order to have cainjector ignore secrets that are known to never be referenced.

### Kind

 feature

### Release Note

<!--

Should we mention this PR in release notes? If so, replace "NONE" with a line of text explaining what changed!

For more details, see: https://git.k8s.io/community/contributors/guide/release-notes.md

-->

```release-note
Add support for restricting the secrets watch list in cainjector
```

irbekrm commented:
Hi @aubm thank you for the PR! 

Agree that cainjector (and also cert-manager controller) watching and caching all `Secret`s is problematic.

We've been thinking about fixing this by making them cache metadata only if possible.

I'm not opposed to also having a flag for cainjector for this if it seems like the metadata approach alone is not enough of an improvement or if it is not feasible for some reason.

The main issue with this PR is that setting this flag will by default break cert-manager's webhook because cainjector is used to sync CA cert from the webhook's self-signed CA cert stored in an (unlabelled) `Secret` in `cert-manager` namespace to the validating and mutating webhook configurations. That would probably need to be solved somehow if we do want this flag.

aubm commented:
Hello @irbekrm, thank you for your feedback!

> We've been thinking about fixing this by making them cache metadata only if possible.

Let me know if there is something I can do to help with that :)

In the meantime, what would you think if I reversed the logic in this PR? We could have the flag ignore secrets with a given set of labels instead.

irbekrm commented:
I'll take a look this week at whether the metadata caching is something that can be implemented in relatively straightforward way and then we can have a discussion about what approach can be realistically taken to solve this soon-ish. If possible I would rather not have the reverse flag as it would require folks to have to set specific labels on cert-manager unrelated resources for cert-manager to not process them - that doesn't seem like a good pattern. But similarly also am aware that this issue needs to be solved.

aubm commented:
Hello @irbekrm 👋

Did you have time to look into this?
I could look to make myself available if you need a hand, don't hesitate to poke!

Cheers,

aubm commented:
Hey @irbekrm, just to let you know that I just updated the PR and changed the new option to restrict the secrets watch list based on a field selector instead of a label selector.

I think that should address the issue your raised here:

> The main issue with this PR is that setting this flag will by default break cert-manager's webhook because cainjector is used to sync CA cert from the webhook's self-signed CA cert stored in an (unlabelled) Secret in cert-manager namespace to the validating and mutating webhook configurations. That would probably need to be solved somehow if we do want this flag.

Let me know what you think!

SgtCoDFish commented:
/kind feature
/ok-to-test
